src/handlers/pull-reviewer.ts

```diff
@@ -0,0 +1,283 @@
+  }
+
+  /**
+   * Handle the code review process for a pull request
+   * @returns CallbackResult indicating the status and reason
+   */
+  private async _handleCodeReview(): Promise<CallbackResult> {
+    const { payload } = this.context;
+    const pullReviewData = await this.reviewPull();
+    const { reviewComment, confidenceThreshold } = this.parsePullReviewData(pullReviewData.answer);
+
+    if (confidenceThreshold < 0.5) {
+      await this.convertPullToDraft(payload.pull_request.node_id, this.context.octokit);
+    }
+
+    await this.submitCodeReview(reviewComment, confidenceThreshold > 0.5 ? "COMMENT" : "REQUEST_CHANGES");
+    return { status: 200, reason: "Success" };
+  }
+
+  /**
+   * Submit a code review for a pull request
+   * @param review - The review comment
+   * @param status - The review status (APPROVE, REQUEST_CHANGES, COMMENT)
+   */
+  async submitCodeReview(review: string, status: CodeReviewStatus): Promise<void> {
+    const { logger, payload } = this.context;
+    const { number, organization, repository, action, sender } = payload;
+    const { owner, name } = repository;
+
+    logger.info(`${organization}/${repository}#${number} - ${action} - ${sender.login} - ${review}`);
+
+    try {
+      const response = await this.context.octokit.rest.pulls.createReview({
+        owner: owner.login,
+        repo: name,
+        pull_number: number,
+        body: review,
+        event: status,
+      });
+      logger.info(`Code review submitted: ${response.data.html_url}`);
+    } catch (er) {
+      throw this.context.logger.error("Failed to submit code review", { err: er });
+    }
+  }
+
+  /**
+   * Check if a review can be performed on the pull request
+   * @returns boolean indicating if review can be performed
+   */
+  async canPerformReview(): Promise<boolean> {
+    const { logger, payload } = this.context;
+    const { number, repository, action } = payload;
+    const { owner, name } = repository;
+
+    logger.info(`${repository.owner.login}/${repository.name}#${number} - ${action}`);
+
+    // Use octokit.paginate to automatically handle pagination
+    const timeline = await this.context.octokit.paginate(this.context.octokit.rest.issues.listEvents, {
+      owner: owner.login,
+      repo: name,
+      issue_number: number,
+      per_page: 100, // Optional: customize items per page
+    });
+
+    const reviews = timeline.filter((event) => event.event === "reviewed");
+    const botReviews = reviews.filter((review) => review.actor.type === "Bot");
+
+    if (!botReviews.length) {
+      logger.info("No bot reviews found");
+      return true;
+    }
+
+    const lastReview = botReviews[botReviews.length - 1];
+    const lastReviewDate = new Date(lastReview.created_at);
+    const now = new Date();
+    const diff = now.getTime() - lastReviewDate.getTime();
+
+    if (diff < this._oneDay) {
+      throw this.context.logger.error("Only one review per day is allowed");
```

Comment:
Can we get cost estimates on reviews? Is it in the $0.01, $0.1 or $1 range? 

I'm starting to think we should allow more frequent reviews, maybe even once per hour. But as part of a separate task we should consider  implementing specific checks for symptoms of abuse instead of doing a timeout, because I can see legitimate use cases for a good developer shipping fast and actually implementing the changes correctly.  

Also I hope that we have the scenario handled where they have been rate limited but the assignee converted the draft to a finalized pull, but then the bot pushes it back to a draft because the automated review failed to run.

Reply:
it really depends on how big the pull request but for pull requests with around 500 additions/deletions, where the diff lines are relatively large (approximately 100 characters per line), the token count would be roughly 23,000 tokens.

Considering this, processing around 40 reviews could cost approximately $3, which averages out to 10 cents per review at most